src/hypothesis/searchstrategy/strategies.py

```diff
@@ -203,35 +236,70 @@ class OneOfStrategy(SearchStrategy):
     def __init__(self, strategies, bias=None):
         SearchStrategy.__init__(self)
         strategies = tuple(strategies)
-        self.element_strategies = list(strategies)
+        self.original_strategies = list(strategies)
+        self.__element_strategies = None
         self.bias = bias
+        self.__in_branches = False
         if bias is not None:
             assert 0 < bias < 1
             self.sampler = cu.Sampler(
                 [bias ** i for i in range(len(strategies))])
         else:
             self.sampler = None
 
+    def calc_is_empty(self):
+        return len(self.element_strategies) == 0
+
+    @property
+    def element_strategies(self):
+        from hypothesis.strategies import check_strategy
+        if self.__element_strategies is None:
+            strategies = []
+            for arg in self.original_strategies:
+                check_strategy(arg)
+                if not arg.is_empty:
+                    strategies.extend(
+                        [s for s in arg.branches if not s.is_empty])
+            pruned = []
+            seen = set()
+            for s in strategies:
+                if s is self:
+                    continue
+                if s in seen:
```

Comment:
I prefer not to write quadratic code even when n is likely small. Particularly with flattening, it's not necessarily *that* small and it should be able to support arbitrary values.